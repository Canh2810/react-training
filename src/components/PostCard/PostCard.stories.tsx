import { StoryObj, Meta } from '@storybook/react'
import PostCard from '.'
import { POST_MOCK } from '@constants'

export default {
  title: 'Components/PostCard',
  component: PostCard,
} as Meta

type Story = StoryObj<typeof PostCard>

export const Default: Story = {
  args: {
    post: POST_MOCK,
    onClick: () => {},
  },
}
