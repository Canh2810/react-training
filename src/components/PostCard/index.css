.post-card {
  padding: 16px;
  border-radius: 12px;
  border: 1px solid var(--border-color);
  width: fit-content;
  cursor: pointer;
}

.post-card__img {
  width: 354px;
  height: 234px;
  border-radius: 6px;
  margin-bottom: 16px;
}

.post-card__content {
  padding: 8px;
  width: 344px;
  display: flex;
  flex-direction: column;
  gap: 16px;
  align-items: start;
}

.post-card__footer {
  margin-top: 4px;
  display: flex;
  align-items: center;
  gap: 20px;
}

.post-card__author-avatar {
  width: 36px;
  height: 36px;
  border-radius: 50%;
  margin-right: -8px;
}
