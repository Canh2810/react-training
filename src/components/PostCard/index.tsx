// Libs
import { memo } from 'react'

// Types
import { IPost, TypoVariants, commonVariants } from '@types'

// Components
import { Badge, Typography } from '../index'

// Styles
import './index.css'

export interface PostCardProps {
  post: IPost
  onClick: () => void
}

const PostCard = ({ post, onClick }: PostCardProps) => {
  const { title, imageURL, author, category, date } = post || {}
  const { username, avatar } = author || {}

  return (
    <div className="post-card" onClick={onClick}>
      <img className="post-card__img" src={imageURL} alt={title} />
      <div className="post-card__content">
        <Badge variant={commonVariants.Secondary} title={category} />
        <Typography variant={TypoVariants.HeadingLarge}>{title}</Typography>
        <div className="post-card__footer">
          <img
            className="post-card__author-avatar"
            src={avatar}
            alt={username}
          />
          <Typography variant={TypoVariants.TextMedium}>{username}</Typography>
          <Typography variant={TypoVariants.TextMedium}>{date}</Typography>
        </div>
      </div>
    </div>
  )
}

export default memo(PostCard)
