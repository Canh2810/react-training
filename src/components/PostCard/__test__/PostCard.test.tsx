import { fireEvent, render } from '@testing-library/react'
import PostCard from '..'
import { POST_MOCK } from '@constants'

describe('PostCard component', () => {
  const mockClick = jest.fn()

  it('should render correctly', () => {
    const { container } = render(
      <PostCard post={POST_MOCK} onClick={mockClick} />,
    )
    expect(container).toMatchSnapshot()
  })

  it('calls onClick callback when clicked', () => {
    const { container } = render(
      <PostCard post={POST_MOCK} onClick={mockClick} />,
    )
    const postCard = container.querySelector('.post-card')

    if (postCard) {
      fireEvent.click(postCard)
      expect(mockClick).toHaveBeenCalledTimes(1)
    }
  })
})
