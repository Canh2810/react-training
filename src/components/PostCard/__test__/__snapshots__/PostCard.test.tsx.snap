// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard component should render correctly 1`] = `
<div>
  <div
    class="post-card"
  >
    <img
      alt="The Impact of Technology on the Workplace: How Technology is Changing"
      class="post-card__img"
      src="https://images.unsplash.com/photo-1543877087-ebf71fde2be1?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8OXx8bmF0dXJhbHxlbnwwfHwwfHx8MA%3D%3D&auto=format&fit=crop&w=600&q=60"
    />
    <div
      class="post-card__content"
    >
      <span
        class="badge"
        style="background-color: rgba(75, 107, 251, 0.05); color: rgb(75, 107, 251);"
      >
        Technology
      </span>
      <p
        class=""
        style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
      >
        The Impact of Technology on the Workplace: How Technology is Changing
      </p>
      <div
        class="post-card__footer"
      >
        <img
          alt="Tracey Wilson"
          class="post-card__author-avatar"
          src="https://www.w3schools.com/w3images/avatar2.png"
        />
        <p
          class=""
          style="font-family: work-sans; font-size: 18px; font-weight: 400; line-height: 26px; color: rgb(59, 60, 74);"
        >
          Tracey Wilson
        </p>
        <p
          class=""
          style="font-family: work-sans; font-size: 18px; font-weight: 400; line-height: 26px; color: rgb(59, 60, 74);"
        >
          August 20, 2022
        </p>
      </div>
    </div>
  </div>
</div>
`;
